src/ucp/proto/proto_perf.c

```diff
@@ -340,6 +334,96 @@ ucs_status_t ucp_proto_perf_add_func(ucp_proto_perf_t *perf, size_t start,
     return status;
 }
 
+ucs_status_t ucp_proto_perf_aggregate(const char *name,
+                                      const ucp_proto_perf_t *const *perf_elems,
+                                      unsigned num_elems,
+                                      ucp_proto_perf_t **perf_p)
+{
+    const ucp_proto_perf_segment_t *seg;
+    ucp_proto_perf_segment_t *new_seg;
+    ucp_proto_perf_t *perf;
+    ucs_list_link_t **pos;
+    ucs_status_t status;
+    size_t start, end;
+    unsigned i;
+    int valid;
+
+    status = ucp_proto_perf_create(name, &perf);
+    if (status != UCS_OK) {
+        goto err;
+    }
+
+    if (num_elems == 0) {
+        goto out;
+    }
+
+    pos = ucs_alloca(sizeof(*pos) * num_elems);
+    for (i = 0; i < num_elems; ++i) {
+        pos[i] = perf_elems[i]->segments.next;
+    }
+
+    start = 0;
+    for (;;) {
+        end   = SIZE_MAX;
+        valid = 1;
+        for (i = 0; i < num_elems; ++i) {
+            /* Find the next segment that may contain the start point */
+            for (;;) {
+                if (pos[i] == &perf_elems[i]->segments) {
+                    /* One of the elements does not have more segments */
+                    goto out;
+                }
+
+                seg = ucs_container_of(pos[i], ucp_proto_perf_segment_t, list);
+                if (seg->end >= start) {
+                    break;
+                }
+
+                pos[i] = pos[i]->next;
+            }
+
+            if (start >= seg->start) {
+                /* Segment contains the start point */
+                end = ucs_min(end, seg->end);
+            } else {
+                /* Segment does not contain start point, start over from the
+                   beginning of the segment */
+                start = seg->start;
+                valid = 0;
+                break;
+            }
+        }
+
+        if (!valid) {
+            continue;
+        }
+
+        status = ucp_proto_perf_segment_new(perf, start, end, &new_seg);
+        if (status != UCS_OK) {
+            goto err_destroy;
+        }
+
+        ucs_list_add_tail(&perf->segments, &new_seg->list);
+        for (i = 0; i < num_elems; ++i) {
+            seg = ucs_container_of(pos[i], ucp_proto_perf_segment_t, list);
+            ucp_proto_perf_segment_add_funcs(
+                    perf, new_seg, UCS_MASK(UCP_PROTO_PERF_FACTOR_LAST),
```

Comment:
minor: we could have a define for this mask, like `UCP_PROTO_PERF_FACTOR_ALL`

Reply:
it's used only once so I'm not sure it's worth it